src/quicklist.c

```diff
@@ -1242,17 +1242,46 @@ int quicklistDelRange(quicklist *quicklist, const long start,
     return 1;
 }
 
-/* compare between a two entries */
-int quicklistCompare(quicklistEntry* entry, unsigned char *p2, const size_t p2_len) {
+/* Compare a quicklistEntry with a raw value.
+ *
+ * If the entry stores a string (entry->value != NULL), perform a binary-safe
+ * comparison against p2.
+ *
+ * If the entry stores an integer (entry->value == NULL), lazily convert p2 to
+ * a long long using string2ll() once and cache the result using cached_longval
+ * and cached_valid.
+ *
+ * This optimization avoids repeatedly calling string2ll() in tight loops.
+ * - If cached_valid == NULL: skip caching
+ * - If cached_valid == 0: conversion attempted
+ * - If cached_valid == 1/-1: cached result reused
+ *
+ * Returns 1 if equal, 0 otherwise.
+ */
+int quicklistCompare(quicklistEntry *entry, unsigned char *p2, const size_t p2_len,
```

Comment:
why don't we add a new parameter(uint64_t llval) to this function? When the caller knows that it's an integer, he can pass p2 as NULL and pass the integer.
and we can make it the same way for `lpCompare`.